ast/edit/option_editor.go

```diff
@@ -0,0 +1,153 @@
+	"fmt"
+
+	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/semantic"
+	"github.com/influxdata/flux/values"
+)
+
+type EditFn func(opt *ast.OptionStatement) (ast.Expression, error)
+
+/*
+`Edit` passes the `OptionStatement` in the AST rooted at `node` that has the
+specified identifier to `editFn`.
+The function can have side effects on the option statement
+and/or return a non-nil `Expression` that is set as value for the option.
+If the value returned by the edit function is `nil` (or an error is returned) no new value is set
+for the option statement (but any, maybe partial, side effect is applied).
+`Edit` returns whether it could find and edit the option (possibly with errors) or not.
+*/
+func Edit(node ast.Node, optionIdentifier string, editFn EditFn) (bool, error) {
+	oe := &optionEditor{identifier: optionIdentifier, editFn: editFn, err: nil}
+	ast.Walk(oe, node)
+
+	if oe.err != nil {
+		return oe.found, oe.err
+	}
+
+	return oe.found, nil
+}
+
+// Creates an `EditFn` for setting the value of an `OptionStatement`.
+func EditValueFn(expr ast.Expression) EditFn {
+	return func(opt *ast.OptionStatement) (ast.Expression, error) {
+		return expr, nil
+	}
+}
+
+/*
+Creates an `EditFn` for updating the values of an `OptionStatement` that has an
+`ObjectExpression` as value.
+*/
+func EditObjectFn(keyMap map[string]values.Value) EditFn {
+	return func(opt *ast.OptionStatement) (ast.Expression, error) {
+		obj, ok := opt.Assignment.Init.(*ast.ObjectExpression)
+		if !ok {
+			return nil, fmt.Errorf("value is not an object expression, it is %s instead", opt.Assignment.Init.Type())
+		}
+
+		for _, p := range obj.Properties {
+			value, found := keyMap[p.Key.Name]
+			if found {
+				p.Value = GetLiteral(value)
+			}
+		}
+
+		return nil, nil
+	}
+}
+
+func GetLiteral(v values.Value) ast.Expression {
+	var literal ast.Expression
+	switch v.Type().Nature() {
+	case semantic.Bool:
+		literal = &ast.BooleanLiteral{Value: v.Bool()}
+	case semantic.UInt:
+		literal = &ast.UnsignedIntegerLiteral{Value: v.UInt()}
+	case semantic.Int:
+		literal = &ast.IntegerLiteral{Value: v.Int()}
+	case semantic.Float:
+		literal = &ast.FloatLiteral{Value: v.Float()}
+	case semantic.String:
+		literal = &ast.StringLiteral{Value: v.Str()}
+	case semantic.Time:
+		literal = &ast.DateTimeLiteral{Value: v.Time().Time()}
+	case semantic.Duration:
+		literal = &ast.DurationLiteral{
+			Values: []ast.Duration{
+				{
+					Magnitude: int64(v.Duration()),
+					Unit:      "ns",
+				},
+			},
+		}
+	case semantic.Regexp:
+		literal = &ast.RegexpLiteral{Value: v.Regexp()}
+	case semantic.Array:
+		arr := v.Array()
+		arrExpr := &ast.ArrayExpression{Elements: make([]ast.Expression, arr.Len())}
+
+		arr.Range(func(i int, el values.Value) {
+			arrExpr.Elements[i] = GetLiteral(el)
+		})
+
+		literal = arrExpr
+	case semantic.Object:
+		obj := v.Object()
+		objExpr := &ast.ObjectExpression{Properties: make([]*ast.Property, obj.Len())}
+		ip := new(int)
+		*ip = 0
```

Comment:
This logic can be simplified. The `make` function accept three arguments when creating a slice: `make(type, length, capacity)`.

So you can create the Properties slice with enough capacity but zero length, then you can append to the slice which will grow its length, but not is capacity.

Generic example:
```
s := make([]int, 0, len(mymap))
for k ,v := range mymap {
    s = append(s, Pair{k,v})
}
```

This is a very common pattern when copying a map into a slice.

If you are interested in more details on how slices work https://blog.golang.org/go-slices-usage-and-internals